requests/signed_message.go

```diff
@@ -0,0 +1,169 @@
+package requests
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// SignedRequest is a builder for constructing a http.Request with Yoti signing
+type SignedRequest struct {
+	Key        *rsa.PrivateKey
+	HTTPMethod string
+	BaseURL    string
+	Port       int
+	Endpoint   string
+	Headers    map[string][]string
+	Params     map[string]string
+	Body       []byte
+	Error      error
+}
+
+func (msg *SignedRequest) signDigest(digest []byte) (string, error) {
+	hash := sha256.Sum256(digest)
+	signed, err := rsa.SignPKCS1v15(rand.Reader, msg.Key, crypto.SHA256, hash[:])
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(signed), nil
+}
+
+func getTimestamp() string {
+	return strconv.FormatInt(time.Now().Unix()*1000, 10)
+}
+
+func getNonce() (string, error) {
+	nonce := make([]byte, 16)
+	_, err := rand.Read(nonce)
+	return fmt.Sprintf("%X-%X-%X-%X-%X", nonce[0:4], nonce[4:6], nonce[6:8], nonce[8:10], nonce[10:]), err
+}
+
+// WithPemFile loads the private key from a PEM file reader
+func (msg SignedRequest) WithPemFile(in []byte) SignedRequest {
+	block, _ := pem.Decode(in)
+	if block == nil {
+		msg.Error = errors.New("Not PEM-encoded")
+		return msg
+	}
+	if block.Type != "RSA PRIVATE KEY" {
+		msg.Error = errors.New("Not an ESA Private Key")
+		return msg
+	}
+
+	msg.Key, msg.Error = x509.ParsePKCS1PrivateKey(block.Bytes)
+	return msg
+}
+
+// Request builds a http.Request with signature headers
+func (msg SignedRequest) Request() (request *http.Request, err error) {
+	if msg.Error != nil {
+		return nil, msg.Error
+	}
+	// Check for mandatorys
+	if msg.Key == nil {
+		err = fmt.Errorf("Missing Private Key")
+		return
+	}
+	if msg.HTTPMethod == "" {
+		err = fmt.Errorf("Missing HTTPMethod")
+		return
+	}
+	if msg.BaseURL == "" {
+		err = fmt.Errorf("Missing BaseURL")
+		return
+	}
+	if msg.Endpoint == "" {
+		err = fmt.Errorf("Missing Endpoint")
+		return
+	}
+
+	// Mangle BaseURL to allow for optional port number
+	baseURL := msg.BaseURL
```

Comment:
Can an optional port be included in the `BaseURL` and drop the separate `Port` to simplify this?

Reply:
This way makes sandbox creation more 'straightforward' - since creating and using the sandbox happen on different ports and it's not too difficult in Go.

Reply:
Could we use https://golang.org/pkg/net/url/ to parse/validate, add the port, then reassemble?

Reply:
Port is a getter, not a setter. Taking on board what Ed said in standup though.